### PR TITLE
Project Bidding

### PR DESCRIPTION
**Pull request by omar-sarfraz**
- Added route for a freelancer to bid on the project
- Created relationships between models,
    - User --> Project (One to Many)
    - User --> Bid (One to Many)
    - Project --> Bid (One to Many)
- Created a bid submit form in frontend and integrated with backend
- Added necessary validation (Only freelancer can submit a bid)
- Modified the backend route to include bids data with the project (Only for owner)
- Displayed the list of bids to project owner

### CODE REVIEW

**Comment by omar-sarfraz** [comment]
- Added a backend route to accept a bid for client
- Added a middleware to check if the current client is project owner
- Added a button on frontend to accept a bid
- Displayed a badge on accepted bid

File: Backend/controllers/bids/addBid.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { Op } from "sequelize";
+
+import { BidType } from "../../lib/types";
+import bidSchema from "../../validation/Bid";
+import { Bid } from "../../models/Bid";
+import { createJoiError } from "../../lib/utils";
+
+export const addBid = async (req: Request, res: Response) => {
+    const bid: BidType = req.body;
+    const projectId: number = parseInt(req.params.projectId);
```
Comment (larebsyed) [inline Backend/controllers/bids/addBid.ts:11]: please check if the parsed output is `Nan`

File: Backend/controllers/bids/deleteBid.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Request, Response } from "express";
+import { Bid } from "../../models/Bid";
+
+export const deleteBid = async (req: Request, res: Response) => {
+    const bidId: number = parseInt(req.params.bidId);
```
Comment (larebsyed) [inline Backend/controllers/bids/deleteBid.ts:5]: check for `Nan`

File: Backend/controllers/projects/getProjectById.ts [hunk 3]
```diff
@@ -1,18 +1,54 @@
...
 
-const getProjectById = async (req: Request, res: Response) => {
-    const projectId = req.params.id;
+const getProjectById = async (expressReq: Request, res: Response) => {
+    const req = expressReq as CustomRequest;
+    const projectId = req.params.projectId;
 
     try {
-        const project = await Project.findOne({ where: { id: projectId } });
+        let project = await Project.findOne({ where: { id: projectId } });
```
Comment (larebsyed) [inline Backend/controllers/projects/getProjectById.ts:13]: convert this to a single query and filter in backend.

File: Backend/middlewares/isBidOwner.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    const projectId: number = parseInt(customRequest.params.projectId);
+    const bidId: number = parseInt(customRequest.params.bidId);
+    const freelancerId: number | undefined = customRequest.user.id;
+
+    // const existingProject = Project.findOne({
+    //     where: {
+    //         [Op.and]: [{ id: projectId }],
+    //     },
+    // });
```
Comment (larebsyed) [inline Backend/middlewares/isBidOwner.ts:18]: remove this commented code

File: Backend/models/User.ts [hunk 5]
```diff
@@ -1,8 +1,13 @@
 import { sequelize } from "../lib/sequelize";
 import { DataTypes } from "sequelize";
 import { USER_ROLES } from "../lib/utils";
+import { Project } from "./Project";
+import { Bid } from "./Bid";
+import { decrypt } from "@omar-sarfraz/caesar-cipher";
 
-export const User = sequelize.define("User", {
+const key = parseInt(process.env.PORT || "0");
```
Comment (larebsyed) [inline Backend/models/User.ts:8]: fix the env variable

File: Backend/models/User.ts [hunk 6]
```diff
@@ -31,3 +39,11 @@ export const User = sequelize.define("User", {
         allowNull: false,
     },
 });
+
+// User can have many project
+User.hasMany(Project, { foreignKey: { allowNull: false }, onDelete: "CASCADE" });
```
Comment (larebsyed) [inline Backend/models/User.ts:44]: Don't cascade on user delete

File: Backend/models/User.ts [hunk 7]
```diff
@@ -31,3 +39,11 @@ export const User = sequelize.define("User", {
...
+Project.belongsTo(User);
+
+// User can have many bids
+User.hasMany(Bid, { foreignKey: { allowNull: false }, onDelete: "CASCADE" });
```
Comment (larebsyed) [inline Backend/models/User.ts:48]: Don't cascade on user delete

File: Backend/validation/Bid.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import Joi from "joi";
+
+const bidSchema = Joi.object({
```
Comment (larebsyed) [inline Backend/validation/Bid.ts:3]: move it to controller

File: React Project/src/components/BidsList.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    canAccept,
+    acceptedBid,
+    projectId,
+    userToken,
+    setProject,
+}: PropType) {
+    const { toast } = useToast();
+    const navigate = useNavigate();
+
+    const handleAcceptBid = async (id: number) => {
```
Comment (larebsyed) [inline React Project/src/components/BidsList.tsx:29]: move the functions in separate files.

File: React Project/src/components/Description.tsx [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+const MAX_CHAR_LENGTH = 300;
+
+export default function Description({
+    description,
+    customClasses,
+}: {
+    description: string;
+    customClasses?: string;
+}) {
+    const [open, setOpen] = useState<boolean>(false);
```
Comment (larebsyed) [inline React Project/src/components/Description.tsx:13]: remove type `<boolean>`

**Review by larebsyed — APPROVED** [review]
Minor changes requested

**Review by larebsyed — APPROVED** [review]

**Pull request merged**